main/webapp/modules/core/scripts/dialogs/add-rows-dialog.js

```diff
@@ -0,0 +1,101 @@
+const AddRowsDialog = (function (path) {
+  let level;
+  const $dialog = $(DOM.loadHTML("core", path));
+
+  class Row {
+    #isFlagged;
+    #isStarred;
+    #cells;
+
+    constructor(isFlagged= false, isStarred= false, cells = []) {
+      this.#isFlagged = isFlagged;
+      this.#isStarred = isStarred;
+      this.#cells = cells;
+
+      return this;
+    }
+
+    serialize() {
+      return JSON.stringify({
+        flagged: this.#isFlagged,
+        starred: this.#isStarred,
+        cells: this.#cells
+      });
+    }
+  }
+
+  return {
+    prependBlankRow: () => _submit([ new Row().serialize() ], 0),
+    appendBlankRow: () => _submit([ new Row().serialize() ], theProject.rowModel.total),
+    initDialog: initDialog,
+  };
+
+
+  /**
+   * Initialize and open dialog
+   */
+  function initDialog() {
+    const $elements = DOM.bind($dialog);
+
+    $elements["dialogHeader"].html($.i18n("core-views/add-rows"));
+    $elements["addRowsCountLabel"].html($.i18n("core-views/add-rows/count-label"));
+    $elements["addRowsPositionLabel"].html($.i18n("core-views/add-rows/insertion-label"));
+
+    $elements["optionBeginning"].val(0);
+    $elements["optionEnd"].val(theProject.rowModel.total);
+
+    $elements["cancelButton"].html($.i18n('core-buttons/cancel'));
+    $elements["cancelButton"].on('click', _dismissDialog);
+
+    $elements["okButton"].html($.i18n('core-buttons/ok'));
+
+    $elements["dialogForm"].on('submit', function(event) {
+      event.preventDefault();
+      const $form = $(this);
+      let index = $form.find("select#add-rows-position").val();
+
+      const rowCount = parseInt($form.find("input#add-rows-count").val());
+      const data = Array(rowCount).fill(new Row().serialize());
+
+      _submit(data, index, _dismissDialog);
+    });
+
+    level = DialogSystem.showDialog($dialog);
+
+  } // end initDialog
+
+  /**
+   * Remove dialog
+   * @private
+   */
+  function _dismissDialog() {
+      DialogSystem.dismissUntil(level - 1);
+  }
+
+  /**
+   * Send POST request to the add rows command
+   * @param {array} data: Serialized row data
+   * @param {number} index: the project index into which new rows are inserted
+   * @param {function} onDone: a callback function that fires after the data table view has updated
+   * @private
+   */
+  function _submit(data, index, onDone = null) {
+    return Refine.postCoreProcess(
+      "add-rows",
+      null,
+      {
+        "rows[]": data,
+        "index": index,
+      },
+      null,
+      {
+        "onDone": (o) => Refine.fetchRows(
+          theProject["rowModel"]["start"],
+          null,
+          () => ui.dataTableView.update(onDone),
+          null),
```

Comment:
Instead of manually triggering the update of the UI, I think it would be simpler to use the existing utilities we have to signal what the operation changed, and let the core system schedule those updates. This means you shouldn't need an `onDone` callback but simply pass flags to indicate which parts of the UI should be refreshed, like here:
https://github.com/OpenRefine/OpenRefine/blob/183246f6b0fde965d4d2d7bdcf6595dc97e857a9/main/webapp/modules/core/scripts/dialogs/column-reordering-dialog.js#L95
or here:
https://github.com/OpenRefine/OpenRefine/blob/183246f6b0fde965d4d2d7bdcf6595dc97e857a9/main/webapp/modules/core/scripts/views/data-table/menu-reconcile.js#L399

In this case I think `modelsChanged: true` should be necessary because you are changing the number of rows in the project. It should be sufficient because that triggers an update of basically everything.

That should have the effect of updating other parts of the UI: for instance, the facets.

Reply:
Awesome, I wish I knew about this sooner. Are these existing utilities documented anywhere? I really struggled figuring out how to maintain state between the client and server when writing JavaScript. I'll also write a Cypress test for an update to facets.